### ISSUE TITLE
[1.11.0] IPv6 addresses not available for a brief period after container start

### ISSUE THREAD

**Issue by thaJeztah** [docker/docker#22157]
(opening a separate issue for tracking in the 1.11.x milestone)

As reported in https://github.com/docker/docker/issues/18871#issuecomment-211974305

/cc @narqo @aboch

It feels that this behaviour was reverted in docker 1.11, isn't it?

~~~
  › sudo docker version
Client:
 Version:      1.11.0-rc5
 API version:  1.23
 Go version:   go1.5.3
 Git commit:   6178547
 Built:        Mon Apr 11 21:07:24 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.0-rc5
 API version:  1.23
 Go version:   go1.5.3
 Git commit:   6178547
 Built:        Mon Apr 11 21:07:24 2016
 OS/Arch:      linux/amd64

  › sudo docker run --rm -i debian ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
20: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 2001:db8:1::242:ac11:2/64 scope global tentative
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:2/64 scope link tentative
       valid_lft forever preferred_lft forever
~~~

**Comment by aboch** [comment]
@narqo 
I cannot reproduce with the released 1.11. Is it consistent for you ?

```
$ docker version | grep 'Vers\|Git'
 Version:      1.11.0
 Git commit:   4dc5990
 Version:      1.11.0
 Git commit:   4dc5990
$ docker run --rm debian ip a sh eth0
280: eth0@if281: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 2001:db8:abcd::242:ac11:2/64 scope global nodad  <-- <--
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:2/64 scope link tentative 
       valid_lft forever preferred_lft forever
$ 
```

Also, can you share your kernel's details.

**Comment by narqo** [comment]
@aboch I've updated to the latest docker-engine 1.11 but still get "tentative" iface on container starts

~~~
  › sudo docker version | grep 'Vers\|Git'
 Version:      1.11.0
 Git commit:   4dc5990
 Version:      1.11.0
 Git commit:   4dc5990

  › sudo docker run --rm debian ip a sh eth0
26: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 2001:db8:1::242:ac11:2/64 scope global tentative
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:2/64 scope link tentative
       valid_lft forever preferred_lft forever

  › uname -a
Linux var01f.vsfront.yandex.net 3.13.0-35-generic #62-Ubuntu SMP Fri Aug 15 01:58:42 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
~~~

I use IPv6-only linux machine in our inhouse OpenStack cloud.

**Comment by aboch** [comment]
@narqo Thanks for confirming and providing the extra details.

The [code](https://github.com/docker/docker/blob/v1.11.0/vendor/src/github.com/docker/libnetwork/osl/interface_linux.go#L358) which programs the IPv6 address with the NO_DAD flag has not changed since it was added.

It use the vishvananda/netlink library to program the IPv6 address with the NO_DAD flag. Seeing you are on a 3.13 kernel, it makes me guess whether the nodad flag can be programmed.

Can you check if you can program an IPv6 address with `ip addr add` and the `nodad` option on some interface on your host ?

**Comment by narqo** [comment]
> It use the vishvananda/netlink library to program the IPv6 address with the NO_DAD flag. Seeing you are on a 3.13 kernel, it makes me guess whether the nodad flag can be programmed.

Looks like you're right and the problem is with my kernel. I've checked this on a clean virtual machine in the cloud (same configuration as one I've showed above), but with 3.19 kernel (e.g. `linux-image-generic-lts-vivid` Ubuntu package) and everything works as expected:

~~~
  › uname -r
3.19.0-58-generic

  › sudo docker run --rm -i debian ip a s eth0
12: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 2001:db8:1::242:ac11:2/64 scope global nodad   ←
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:2/64 scope link tentative
       valid_lft forever preferred_lft forever
~~~

**Comment by aboch** [comment]
Thanks @narqo for confirming.
We still need to see if it is a 3.13 limitation or a bug in the go netlink library w/ 3.13.

If you do the following on your 3.13 machine, is the nodad shown in the o/p ?

`ip link add iface type dummy; ip addr add 2001:db8:abcd::32/63 dev iface nodad; ip link set iface up; ip addr show iface | grep global`

**Comment by mpalmer** [comment]
The `nodad` flag has been around since the beginning of time (well, IPv6 time, anyway).  It definitely works on 3.13:

```
$ uname -r; ip -6 ad sh |grep nodad
3.13.0-45-generic
    inet6 2001:db8::42/64 scope global nodad
```

That being said, I too can reproduce the bug on a scratch VM running 3.13:

```
root@scratch:~# uname -r; docker run --rm alpine ip ad sh             
3.13.0-68-generic
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN 
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
16: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state DOWN 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 2001:db8::242:ac11:2/64 scope global tentative 
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:2/64 scope link tentative 
       valid_lft forever preferred_lft forever
```

While upgrading that exact same system to a 3.19 kernel makes the problem mysteriously disappear.

**Comment by aboch** [comment]
@mpalmer 
Yeah, I also verified manually creating and configuring a v6 address with nodad on a 3.13 works fine.
While with docker run would fail. Given docker completely depends on vishvananda/netlink pkg to do these operations, I will open an issue there.

I doubt this can be fixed in time for 1.11.1 though.

**Comment by mpalmer** [comment]
Release note time!  At least the problem can be worked around by upgrading kernel, but it's a right weird problem.

**Comment by mlaventure** [comment]
Moving this to 1.12.0 since there's a nice workaround and it won't be fixed in time anyway.

**Comment by narqo** [comment]
> since there's a nice workaround

By "nice" you mean an ability to upgrade to 3.19 kernel? ;)

**Comment by thaJeztah** [comment]
ping @aboch I see https://github.com/vishvananda/netlink/pull/121 was merged; are those changes in 1.12, and should they resolve this issue for kernel 3.13?

**Comment by aboch** [comment]
@thaJeztah That's correct.

**Comment by thaJeztah** [comment]
Thanks! I'll close this issue as it should be resolved in 1.12, but let us know if you're still able to reproduce
